### PR TITLE
Added JSONObject.expand()

### PR DESCRIPTION
**Pull request by fossterer**
**What problem does this code solve?**
- Adds _JSONObject.expand()_ implementation
- **Reference:** https://blog.fossterer.com/Technical/contributing-jsonutils-getexpandedjson-org-json.html

**Risks**
Low

**Changes to the API?**
Adds new

**Will this require a new release?**
No

**Should the documentation be updated?**
Yes

**Does it break the unit tests?**
No

**Was any code refactored in this commit?**
No

**Review status**

### CODE REVIEW

File: src/main/java/org/json/JSONUtils.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	 * 
+	 * <pre>
+	 * {		{		{
+	 *   a: b,	  a: b,	  	  a: b,
+	 *   c: d1,	  c: d2,  	  c: d3,
+	 *   e: f1	  e: f2	  	  e: null
+	 * }		}		}
+	 * </pre>
+	 */
+	public static List<JSONObject> getExpandedJSON(JSONObject incomingJSON) {
```
Comment (fossterer) [inline src/main/java/org/json/JSONUtils.java:108]: You may start reading from here.

**Comment by fossterer** [comment]
@stleary You may start reading from [Line 108](https://github.com/stleary/JSON-java/pull/585/files#diff-d37b32c2eb8e100341e9cf7ea26352950c70874d5071112abaa81e28d98fccd4R108)

The rest are tests and test resources

**Comment by johnjaylward** [comment]
When I think of a "utility" class, I generally think of 2 use cases.

1. A class with commonly used functions that are not part of the existing API but would be useful under multiple use-cases or across multiple domains.
2. A class with functions in it that don't really fit anywhere else because their scope of use is too narrow to fit into the above "general use" class, however they may need to be used in a few key, distinct points of an application.

In general, I highly dislike utility classes of any kind as they feel like an anti-pattern (Why wasn't the function just added to the API in the first place). The exception being when creating a utility class to work with APIs that are not managed by your code.

An instance of a good use of a utility class is something like `StringUtil` where you are adding helping functions to a class that you can not modify.

I've never personally felt the need for the functions like the one this PR uses, and I've never seen a request for the feature here before. While it looks like a lot of effort and good testing went into the feature, this feels more like the second use of "utility" above. But again, why is this a "utility class" instead of just being part of the API?

Maybe I'm wrong, and this is a feature hotly wanted, but my gut based on the history of this project tells me that this is more of a special case and doesn't fit the "general use case" that this library tries to adhere to.

Now, our JSONObject and JSONArray classes certainly fit into "God Class" anti-pattern as there are definitely functions in there that would be more appropriate to be elsewhere (in the tokener for instance, or perhaps a package-visible class). However, given the current design, If these new functions were to be accepted, I'd probably look into making these instance methods on their respective classes instead. i.e. Move `JSONUtils.getExpandedJSON(JSONObject incomingJSON)` to `JSONObject.expand()` and move `JSONUtils.getExpandedJSON(List<JSONObject> incomingJSONList)` to be a static function on `JSONObject` or just drop it altogether since it's no more than a wrapper around the first function.

**Comment by fossterer** [comment]
Hello,

Just notifying that I attended to the suggestions given above. All changes have been pushed

**Comment by stleary** [comment]
Hi @fossterer, Agreed, nice coding and tests, but in order to progress this PR, the description needs to include text explaining how this feature is used and how it aligns with the project goals. I can't come up with a use case.  If it only serves some purpose in your project, then it should probably be implemented as part of your project.

**Comment by fossterer** [comment]
Glad to have your comments @stleary. I'm a fan of this nimble library and always looked forward to give back something to improve.
As I solved this challenge I faced one day in a generic form so close to underlying library (`JSONObject`) I took that as an opportunity to contribute.

So this project may continue to be light-weight, this code may not be welcome here. Do any JSONObject-friendly projects come to your mind that can be a new home for this implementation? Please respond and then you may close this PR.

Thank you

**Comment by stleary** [comment]
@fossterer  Although expand() was not the right match, issues with the label "Coding Issue" need someone to fix/enhance the code:
#563 
#561 
#551
The JSONPointer issues are interesting because the original author is not available and no one else has come up to speed yet. Consider taking on one of these issues if you want.

**Pull request closed without merging**